operands_and_partials refactor

seantalts opened the pull request:
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Replace the existing OperandsAndPartials with a new one that attempts to be a little less confusing (YOU be the judge!). It also adds multivariate support.

#### Documentation:

#### Reviewer Suggestions: 
I had a few questions and wasn't sure how to do some things throughout but tried to add comments in the code with a TODO or something similar to catch these situations. That said, the code seems to work on all the tests that I can run in a short amount of time (will use Jenkins to run the full distribution tests, etc).

I also haven't added any doc yet; I plan to doc the main `operands_and_partials` class while the tests are running on Jenkins (but probably not the stuff I've added in the `detail` namespace).

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Dual Space LLC

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

seantalts commented:
Let me fix the tests before anyone wastes time on this!

syclik commented:
@seantalts, I'll take a look at this over this week.

Quick question: can we come up with a better name?

syclik commented:
oh yeah... you think it's time for a review?

seantalts commented:
So I sat with Bob today went over the design for a while and got some changes to make. I'll hopefully finish those up tomorrow - I'll post back here and tag you when I'm done? Or it might make sense to just have @bob-carpenter do the review since he's fairly steeped in the new design at this point.

syclik commented:
Great! Please ping me anyway. I'd like to see the changes and having an
extra set of eyes never hurt on something like this.

On Mon, May 8, 2017 at 10:58 PM, seantalts <notifications@github.com> wrote:

> So I sat with Bob today went over the design for a while and got some
> changes to make. I'll hopefully finish those up tomorrow - I'll post back
> here and tag you when I'm done? Or it might make sense to just have
> @bob-carpenter <https://github.com/bob-carpenter> do the review since
> he's fairly steeped in the new design at this point.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/547#issuecomment-300047848>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_Fw0ZPFkpFe8v1AZJaDvthkyXoeEbks5r39ZFgaJpZM4NQAv4>
> .
>

seantalts commented:
Will do!

seantalts commented:
@syclik okay, addressed Bob's comments and it's passing my smoke tests. Feel free to take a look (or wait until it passes Jenkins just in case). Thanks!

bob-carpenter commented:
I can review this later this week.  I'm  a bit pressed now on proposals and admin, but should be able to do it later this week.

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+#define STAN_MATH_PRIM_SCAL_META_OPERANDS_AND_PARTIALS_HPP
+
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/meta/broadcast_array.hpp>
```
alphabetize

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+
+        ops_partials_edge() {}
+        explicit ops_partials_edge(const Op& /* a */) {}
+        //
```
remove

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+        int size() const { return 0; }
+      };
+
+      // Base class shared between fvar and var specializations of uni- and
```
Any reason this isn't a doxygen doc?

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+      detail::ops_partials_edge<ViewElt, Op3> edge3_;
+      detail::ops_partials_edge<ViewElt, Op4> edge4_;
+    };
+  }  // end namespace math
```
I prefer not having `// end namespace`, but that's really a style thing

seantalts replied:
Ah right, in the Google Style Guide it's just `  // namespace math`. Re: style I think the closer we stick to some external reference point the better; easier to onboard people and less exceptions to remember (and less tooling required to deal with the exceptions). Plus eliminates a whole class of bike shedding.

syclik commented on stan/math/prim/scal/meta/broadcast_array.hpp:
```diff
@@ -0,0 +1,34 @@
+namespace stan {
+  namespace math {
+    namespace detail {
+      template <typename T>
```
add doxygen doc

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+      public:
+        const Op& operand_;
+        broadcast_array<ViewElt> partials_;
+        explicit ops_partials_edge_single(const Op& a)
```
rename the argument `a` to `operand` or `op`

seantalts replied:
Will do.

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+                            const Op4& op4) {}
+      typedef double ViewElt;
+
+      T_return_type build(ViewElt value) {
```
Doc function

syclik commented on test/unit/math/prim/scal/meta/operands_and_partials_test.cpp:
```diff
@@ -0,0 +1,14 @@
+  operands_and_partials<double> o1(0.0);
+  operands_and_partials<double,double,double,double> o2(0.0, 0.0, 0.0, 0.0);
+
+  // TODO(Sean): Learn why it's size 6
```
What's size 6?

seantalts replied:
errant comment!

syclik commented on test/unit/math/prim/scal/meta/operands_and_partials_test.cpp:
```diff
@@ -0,0 +1,14 @@
+  EXPECT_GT(sizeof(&o2), sizeof(o2));
+
+  SUCCEED() << "Construction should be ok.";
+}
```
test the `build()` method?

seantalts replied:
Will do.

syclik commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,89 @@
+  namespace math {
+    namespace detail {
+      template <typename ViewElt, typename Op>
+      struct ops_partials_edge {
```
Doc class including member functions.

syclik commented:
@seantalts, I think a little more doc / tests would go a long way. I got lost with the `ops_partial_*` classes. I'm not sure what `dump_partials` and `dump_operands` methods are supposed to do.

In general, I'm guessing it all works. It looks pretty clean. Just having a little more doc would help. (I'd be willing to help doc if I could only piece it together.)

seantalts commented:
The stuff in the `detail` namespace isn't part of the Math library's public API, not even for people who might be writing distributions. There was a discussion at some point that lead to us deciding that we can safely hide implementation details in there without full doxygen doc. That said, I would like to make everything as easy to understand as the underlying conceptual complexity allows. I'll add a bit of doc that explains how `dump_partials` and `dump_operands` are used and some doc over the edge base template to explain what they're supposed to represent.

w.r.t. tests there are a lot more than there used to be! Though I'm realizing now that I haven't tested forward mode's multivariate implementation and so of course it has a problem:

```
./stan/math/fwd/mat/meta/operands_and_partials.hpp:61:55: error: invalid operands to binary expression ('Scalar' (aka 'stan::math::var') and 'const stan::math::fvar<stan::math::var>')
              derivative += this->partials_vec_[i](j) * this->operands_[i](j);
```

Am I right in thinking we should in fact be able to multiply a `var` and an `fvar<var>` (since I think this is how one calculates the derivative here)?

bob-carpenter commented:
dump_operands is what writes into the underlying vari** for rev, does nothing for double, and deals with collecting daughters in fvar<T>.

dump_partials writes the partials to partials_t* in rev, does nothing for double, and dumps out the tangents for fvar<T>.

I can also help doc.  I think I follow all the code now after Sean walked me through it.

bob-carpenter commented:
No, we shouldn't be able to multiply a var and fvar<var>.  Where is that happening?   You should be multiplying values by tangents, so it'll all be var on the inside of fvar<var>, not mixed.

seantalts commented:
Woops! I was missing a `.d_` in the forward multivariate `build` method. Fixed that and added a test, though I don't really understand what's going on in the test right now and might need help working out what the gradients should be in these multivariate cases.

bob-carpenter commented:
What's multivariate?  Did you mean higher order?

seantalts commented:
Forward multivariate meaning dealing with operands that are something like `std::vector<Eigen::Matrix<fvar<double>, -1, -1> >`.

bob-carpenter commented:
Packing variables into containers doesn't change the gradients.

That is, if you have 

  f(x, y) = g(x, y)

vs.

  f(real[] z) = g(z[1], z[2])

it's the same derivatives with respect to x and y as w.r.t. z[1] and z[2].

seantalts commented:
@syclik wanted to see some performance numbers. Here's the results of `stan/src/tests/performance`:
my ops and partials:
[       OK ] performance.run (134949 ms)
[       OK ] performance.run (128901 ms)
[       OK ] performance.run (119232 ms)


old one:
[       OK ] performance.run (124822 ms)
[       OK ] performance.run (123101 ms)
[       OK ] performance.run (125092 ms)

Seems to be about the same to me.

bob-carpenter commented:
What's that measuring?

just operands_and_partials?  if so, how are derivative times measured?

just reverse mode or also double and fwd?

seantalts commented:
It's this: https://github.com/stan-dev/stan/blob/develop/src/test/performance/logistic_test.cpp

@syclik told me about it yesterday.

I'm not sure why there's so much variance - maybe because I was using my computer lightly while the tests were running? Either way, seems close enough.

seantalts commented:
@syclik I added some tests for `zero_vec_or_mat`, which was plausibly stand-alone and not as much of a coupled implementation detail of `operands_and_partials` as the edge specializations are. I also went over forward multivariate and my test with Bob and got that ironed out. I think I also addressed the rest of your comments if you want to take another look.

Many thanks to both of you!

syclik commented on stan/math/prim/mat/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,95 @@
+          }
+        }
+        int size() {
+          return this->operands_.size() * this->operands_[0].size();
```
BUG: This relies on `operands_` to have a size greater or equal to 1. I'm not sure how this is going to behave when `operands_` is size 0.

seantalts replied:
When does that happen?

syclik replied:
It happens when `ops` is 0 length. That could be an `Eigen::Matrix<var, -1, -1>` with either rows == 0 or columns == 0.

syclik replied:
Instead of fixing the code, please doc that this will break if `operands_` is of size 0. In our current use cases, this isn't possible, but I can imagine use cases in the future where this will be a problem.

bob-carpenter replied:
Can we just use the following to avoid the problem?

```
return this->operands_.size() == 0 ? 0 : this->operands_.size() * this->operands_[0].size();
```

I agree that this will never happen with our current usage.  I don't think the switch overhead will be that high here.  We could even go with 

```
if (expected(this->operands_.size() > 0)
  return this->operands_.size() * this->operands_[0].size();
else
  return 0;
```

seantalts replied:
I did basically this (though I think it was called `unlikely`).

syclik commented:
Here are tests that should be added (if we're saying the API boundary is operands and partials):

- test/unit/math/prim/scal/meta/operands_and_partials_test.cpp
  - each `edge*_.partials[0]` throws.
  - each `edge*_.size() == 0`
  - Each `edge*_.dx()` returns 0.0 and is of type `double`
  - Check that `dump_partials()` and `dump_operands()` don't crash.
- test/unit/math/mat/scal/meta/operands_and_partials_test.cpp
  - Construct! Missing constructors of operands_and_partials
  - each `edge*_.partials[0]` throws.
  - each `edge*_.size() == 0`
  - Each `edge*_.dx()` returns 0.0 and is of type `double`
  - Check that `dump_partials()` and `dump_operands()` don't crash.
  - Test with size 0 matrix / arrays.
- test/unit/math/rev/scal/meta/operands_and_partials_test.cpp
  - Missing test! Completely missing the instantiations.
  - Construct each.
  - Verify each `edge*_.partials[0..N]` works / throws appropriately
  - Each `edge*_.size() == 1` or `0`
  - Each `edge*_.dx()`
  - Check that the `vars` constructed through the `ops_partials_edge_single` is put on the correct stack.
- test/unit/math/rev/mat/meta/operands_and_partials_test.cpp
  - Do we test for access of `partials_` past the size? This may not be a test; it'd be user error, but we should doc that it is.
  - Test `edge*_.dump_partials`
  - Test `edge*_.dump_operands`
  - Test `edge*_.dx()`
  - Test `edge*_.size()`
  - Test instantiations of every combination: double, var, std::vector<double>, std::vector<var>, Eigen::Matrix<double>, Eigen::Matrix<var>
  - Also test that passing the same `var`s to different arguments does the right thing with reusing the `var`

That was just for the `prim` and `rev`, and I didn't actually list all the combinatorics since it would take forever. It's tough when you have a complex object exposed and you're not trying to test the individual units of it directly.

seantalts commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,120 @@
+       * @param value the return value of the function we are compressing
+       * @return the value with its derivative
+       */
+      T_return_type build(ViewElt value) {
```
This is funny; Bob actually had me change it in the opposite way. I think there should be a limit of 1 person per code review or else you get conflicting signals! :P

seantalts commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,120 @@
+      operands_and_partials(const Op1& op1, const Op2& op2, const Op3& op3) {}
+      operands_and_partials(const Op1& op1, const Op2& op2, const Op3& op3,
+                            const Op4& op4) {}
+      typedef double ViewElt;
```
Bob had me do this for clarity

syclik replied:
Ha. The reason I'd want it to be private is since we're using a concept and not object inheritance, whoever is looking at this as the concept may mistake `ViewElt` as being part of the concept. It's not (specializations don't actually specify it).

seantalts commented:
* All of the internal methods (`dx`, `dump_partials`, `dump_operands`) are implementation details - should I make some attempt to hide them from the API? What would you recommend? Is this where that `friend` declaration I see around could be useful?
* the throwing is tested in broadcast array's tests
* var construction is tested in other unit tests
* I don't know what you mean by "Missing test! Completely missing the instantiations" - I see instantiations aplenty 

I am curious about this "reusing the var" thing - can you tell me more?

I don't want to spend a lot more time on this as it's extremely low-maginal ROI at this point; Michael actually asked me to stop working on it weeks ago. If there's something quick I can do to get it into a state you like, let me know - otherwise feel free to accept or reject as you like.

seantalts commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,120 @@
+        return value;
+      }
+
+      detail::ops_partials_edge<ViewElt, Op1> edge1_;
```
clarity

syclik replied:
assumption was `ViewElt` would be removed from the struct. Then it would make sense that it's `double` instead of `ViewElt`.

syclik commented:
I thought you were waiting on me to review it. I'll just quickly go through the comments and I'll let you decide on what you want to do.

syclik commented:
> I am curious about this "reusing the var" thing - can you tell me more?

I'm referring to when the same vars are passed in as ops for multiple arguments. You might see it through the Stan language when you have something like `y ~ gamma(theta, theta);`.


> All of the internal methods `(dx, dump_partials, dump_operands)` are implementation details - should I make some attempt to hide them from the API? What would you recommend? Is this where that friend declaration I see around could be useful?

If I were to redesign it, I would not have those methods be part of the object that is exposed through `operands_and_partials`, that is, I wouldn't have it be part of the `operands_and_partials.edge1_` class. Instead, I'd have a struct that accepted `edge1_` on construction and that would then expose `dx, dump_partials, dump_operands`. That object would be used by `operands_and_partials` internally and it wouldn't be exposed to the user. 

> I don't know what you mean by "Missing test! Completely missing the instantiations" - I see instantiations aplenty

First, I missed a test file, so it's not completely missing. But it's missing a lot of instantiations.

This has to do with what we define as the API boundary and how we test it. If we wanted to treat how someone using `operands_and_partials` as the API boundary, then this is missing a lot of instantiations. For it to be tested at the API boundary, you'd really have to see, for each of the edges: `double`, `std::vector<double>`, `Matrix<double, -1, -1>`, `Matrix<double, -1, 1>`, `Matrix<double, 1, -1>`, and then replace all of the doubles with `var`. That's 10 instantiations per edge... you'd really need 4^10 = 1M instantiations to cover it all. The tests there only cover the first edge and not the second, third, and fourth edge.

This is where I think testing the internals cuts down on the testing. If we tested the internal class with the 10 different instantiations, then the burden on testing `operands_and_partials` is cut down to making sure the pieces are instantiated properly and the output is constructed properly.

But please don't think I'm suggesting we do all that. Last we discussed over discourse (or in person) I said I'd review the code from the point of view of testing the API boundary instead of testing the pieces.

Regarding this PR, I don't know what to think. It looks better than what was originally there. I'm happy for it to get in. What are your thoughts on it? I don't want to put more time into it if you're abandoning it.

seantalts commented:
So when you call `y ~ gamma(theta, theta);` where are the `var`s expected to be re-used? I think all of the edge classes take constant references to the `operands` they hold and so from that perspective I think they are re-used? Are you also thinking that in `dump_operands` it should always check that it hasn't already dumped the `vari`s it will try to write out? I could see that being a concern, though I don't think the existing implementation deals with that either. Would it be enough to check if the `operands` references were pointer-equivalent or would we need to store a set of all `vari`s written out so far and check that on every write?

The reason the edge classes are they way they are was to encapsulate the template metaprogramming logic that specializes on all the different types of operands that we accept into one place - separate classes for dump_operands et al would require duplicating that logic, as far as I understand. I agree that it's not ideal that those methods are exposed and look like part of the API. I'll try to think of some other way to deal with that - would welcome more brainstorming and suggestions.

There's no way it's realistic to test all 4^10 possible combinations here. I've chosen a sampling that I think realistically tests most of the branches in the code (though it'd be really nice to have a code coverage plugin right about now). That said, the probability distribution tests already test all of the instantiations we've decided we care about, from what I understand. So I don't think we need to retest that here as well.

I'm happy to spend another half day or day on this but want to timebox it so I can work on other things.

seantalts commented:
re: hiding those internal methods - I think I might be able to make the edges private and expose the `partials` through a weakly templated method like this:
```
template <typename T>
T partials1() { return edge1_.partials_; }
```
...something like that?

bob-carpenter commented:
@seantalts:  the suggestion in https://github.com/stan-dev/math/pull/547#issuecomment-301473832 d would require users to specify `T` --- what options are possible?  Isn't the type of the partials specified at construction time of `operands_and_partials`?

bob-carpenter commented:
@sycklik For https://github.com/stan-dev/math/pull/547#issuecomment-301371344, I like the way @seantalts has it for dumping.

Why in the world would we abandon this?  It's going to let us do the multivariate instantiations better.

I would like to figure out a way to do more thorough tests.  We don't need to test every possible combination if we can test the components then test how they fit together.  I think @syclik was suggesting that, but I don't understand the suggestion.

bob-carpenter commented:
As to https://github.com/stan-dev/math/pull/547#discussion_r116399233, this has happened before.  I'm primarily trying to make suggestions.  I'll try to call out things that have to change, but a lot of this is style.

bob-carpenter commented:
Speaking of style, I think we have to figure out how to set aside issues of what we think is the perfect design in favor of making progress forward.

bob-carpenter commented:
As to https://github.com/stan-dev/math/pull/547#discussion_r116404332, that's not how C++ works.  The base template doesn't tell you anything at all about what's implemented in specializations.  Nor do either have anything to do with concepts used externally.  I think that's a real problem with understanding C++ template designs---they wanted to get some concept representation into the C++17 object language, but they failed.  

The only way to deal with concepts is to document them at the caller's level.  It's possible, for instance, that one specialization of a base template class satisfies concept requirements and another doesn't, and that's fine as long as only the satisfying one is used.

seantalts commented:
Re: https://github.com/stan-dev/math/pull/547#issuecomment-301534591 I think you're right and it likely makes sense to expose the partials type on the edge classes. I'll give that a shot and work on hiding their implementation details.

seantalts commented:
So I can do this kind of thing and use SFINAE to provide these methods only when appropriate (due to univariate o&p having `partials_t` and multivariate having `partials_vec_t`):
```
      typedef detail::ops_partials_edge<ViewElt, Op1> edge1_t;
      template <typename T = edge1_t>
      typename T::partials_vec_t& partials1_vec() { return edge1_.partials_vec_; }
```

Thoughts on this? It also uses a C++11 feature, default function template parameters. Some alternatives:
1. Break out multivariate from univariate
2. leave things as is
3. declare that the edges and the base operands_and_partials classes are `friend`s, and declare those implementation details to be private. 

I think 3 seems like the easiest for right now...

bob-carpenter commented:
This is terrible for threaded commenting as I can't tell what https://github.com/stan-dev/math/pull/547#issuecomment-301607877 is in response to.

What is all this complicated template programming for?

seantalts commented:
I was confused too at first but I think if you click the link it will bring the relevant comment to the top of your page.

In any case, the higher level goal is to hide `dx()`, `dump_partials()`, etc from end users. In the complicated templating post above, I was doing this by hiding the `edges` and just exposing a method to get at their contained partials. The template is used to make sure the `partials1_vec()` method only exists (using SFINAE) when there is an underlying `partials_vec_` member on the `edge1_` member, which happens when it's what I've been calling a "multivariate" edge (i.e. its operands are something like `std::vector<Eigen::Matrix<...>>`

syclik commented:
Please feel free to ignore a bulk of the code-level comments. I was reviewing the pull request with [this conversation](http://discourse.mc-stan.org/t/namespace-stan-math-detail/488/2) from Discourse in mind. My memory is pretty hazy, but I recalled a request (in person) for a deep review at the API boundary, which is what I provided. It was a discussion between documenting and testing internal pieces vs just what's visible as a consumer of the class. Once I started looking at what the probability distribution functions use, it turns out there's a lot more complexity there. The internal classes are exposed directly and not testing those pieces directly makes it really tough to test from the outside.

@seantalts, as I mentioned in an earlier comment before, I'm happy with it going in as-is or possibly with a little work. I'm not expecting everything I mentioned to get fixed (ever). This does leave us in a better place than where we were. I guess the only change I'd want is for the thing I marked as a bug to be made safe.

If there's discrepancy between what I suggest and @bob-carpenter suggests, I'd say go with @bob-carpenter's suggestion.

To be helpful: @seantalts, what could I do to help? Should I put together a list of things that should be finished before merging? Should I add tests? Should I just sit back and let you finish it out?

syclik commented:
@seantalts, although the code is unsafe, it can't be instantiated with the way our probability distributions currently work. I'm going to change the comment to requesting doc instead of fixing the code. There are still ways to construct it so it'll break, but I don't think we can do that with our current use cases.

seantalts commented:
I appreciate the thorough review, sorry for the push back - just didn't want to spend a ton more time on it. I figured out a way to use friendship to limit exposure of those internal methods so we don't have to consider them as part of the public API. 

I will figure out a solution for the BUG above if operands_ is size 0 and change the `detail` to `internal`. Are there any major classes of instantiation you think are missing that I could add quickly?

I also responded to the "var reuse" thing with more questions [here](https://github.com/stan-dev/math/pull/547#issuecomment-301463300), not sure if you saw it? I can try to do something about this as well if you think it's important.

bob-carpenter commented:
For the var reuse thing, I don't think you need to worry about the foo(a, a) case and consolidating the a variables.  We'll just take the hit for the extra overhead.  It'd be hugely complicated to try to reduce this and probably of very limited use (not many distros where you repeat parameters, except the Dirichlet or some symmetric multinomials).

I don't think you need to dream up possible instantiations here.  If we have a sufficient set to implement our distributions, we can always add more specializations later for efficiency.

I haven't been too worried in reviews about carving out a public API because we haven't done it anywhere else.  I am glad you're starting to think about it as I do think it's important.

When we're both back in the office, I'd like to talk about the general testing framework I have in mind that should allow us to blast through 4^10 possibilities.  Whether we always want to do that is another question---if we could somehow set up tests with proofs by induction without having to repeat logic all the way up, that'd be great.  But as is, we're trying to be defensive against downstream ad hoc optimizations.

seantalts commented:
Sounds good, looking forward to it.

bob-carpenter commented:
There's both a `likely()` and an `unlikely()`, so it depends how you want to arrange.  With exceptions, it's usually `unlikely()` in an up-front test.

seantalts commented:
(I think this is ready for review again)

bob-carpenter commented:
Let's go over it tomorrow (Thursday) in person if that's OK.  Then we should be able to resolve any last-minute things and get it merged.

seantalts commented:
Okay, this is ready for another look. I think I addressed all of @bob-carpenter's comments on Thursday - some writing improvements and removing any inheritance from it, simplifying the conceptual burden at the expense of some additional code duplication. I'm pretty happy with it now.

bob-carpenter commented:
Great.  I can take a pass through this Monday.

bob-carpenter commented on stan/math/prim/scal/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,116 @@
+        return value;
+      }
+
+      // These will always be 0 size base template instantiations (above).
```
If they're always zero size, why have them at all?  It's not like they get get inherited by specializations or anything.

seantalts replied:
They are needed to compile - even though the code will never run, the compiler needs to see that edge1_.partials_ exists and is the right type even for arithmetic primitives.

bob-carpenter commented on stan/math/rev/mat/meta/operands_and_partials.hpp:
```diff
@@ -0,0 +1,116 @@
+      template <int R, int C>
+      class ops_partials_edge<double, Eigen::Matrix<var, R, C> > {
+      public:
+        typedef Eigen::Matrix<var, R, C> Op;
```
Don't worry about fixing this, but going forward, I'd much rather see typedefs of the form `op_t` rather than `Op`.  Using `Op` makes it visually ambiguous with a template parameter (and if it were `OP`, with a constant).
